### ISSUE TITLE
Activitypub interop: private posts are not delivered in mastodon if receiver is not mentioned

### ISSUE THREAD

**Issue by git-marijus** [tootsuite/mastodon#7394]
If sending a private message via the "to" or "bto" property set to a subset of followers and those followers are not mentioned, the messages will not be delivered in mastodon.
The mention should probably just trigger notifications but not be mandatory for delivery. At least i did not see anything related in the activitypub spec. 

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nightpool** [comment]
i mentioned this when preparing the impl report but unfortunately didn't have bandwidth to fix it at the time, so it got dropped. https://github.com/tootsuite/mastodon/issues/5631#issuecomment-343041015

should be a pretty easy change if someone wants to take and fix, we can just change it so that mentions are a necessary but not sufficient condition.

**Comment by nightpool** [comment]
yes, because Mastodon's UX has no other way of exposing an "intended
audience" to the user, so it would be very confusing (and a possible spam
vector) if you were notified for a post that apparently had nothing to do
with you.

considering the scope of mastodon as a microblogging application, adding
the additional complex UX for arbitrary audience values is probably not a
desired feature

**Comment by git-marijus** [comment]
> so it would be very confusing (and a possible spam vector) if you were notified for a post that apparently had nothing to do with you

That makes it impossible to share a private message with a subset of followers where one of them should be mentioned. One would have to mention (notify) all of them.

Delivering to "to" and notify the one who is mentioned would be considered less spam i guess. Mastodon could still use mentions for private messages and add the mentions to "to". That would not change anything regarding UX for mastodon but would make interoperability with projects which have a more complex privacy model smoother.

**Comment by nightpool** [comment]
```
Mastodon could still use mentions for private messages and add the mentions to "to". That would not change anything regarding UX for mastodon but would make interoperability with projects which have a more complex privacy model smoother.
```

This is our current behavior, as far as I know. Do you have an example otherwise?

**Comment by nightpool** [comment]
Hmm, unless i'm misunderstanding your proposal..... Are you saying mastodon should do that for posts it *originates*? or posts it *receives*?

**Comment by git-marijus** [comment]
For posts it receives. Sorry if the title and toplevel post are not clear...

**Comment by nightpool** [comment]
> Mastodon could still use mentions for private messages and add the mentions to "to". 

So you're saying if we receive a DM with mentions but no activitypub addressing that includes that person, we should pretend as if they were addressed? How would that help interoperability?

**Comment by nightpool** [comment]
Wouldn't the interoperable thing to do be the *other* way around? create and inject mentions into the status for addresses that are included in activitypub but not mentioned explicitly?

**Comment by ThibG** [comment]
There are two issues with that proposal:
- People mentioned but not in the audience will be able to see the object, which may not be expected by the sender
- This may cause toots to appear in user's timelines without them knowing why (this could easily be fixed by only inserting messages from followed people, though, meaning that if you are in the audience but not mentioned or following the author, you will still be able to view the message if you look for it, but you won't be made aware of its existence in any way)

Either way, this would require some in-depth changes to Mastodon, since it does not stores the audience, only mentions and privacy setting (which is a very rough approximation of the audience).

**Comment by nightpool** [comment]
As I outlined above, requiring both to *and* mention is the only quick fix that's really possible for mastodon. Like I said, allowing users to see a private post where they're not mentioned (and isn't addressed to followers) presents both severe UX and technical hurdles for mastodon.

**Comment by redmatrix** [comment]
Understood this is difficult for Mastodon to implement, but please understand that this is basic ActivityPub communication as described in https://w3c.github.io/activitypub, see specifically EXAMPLE2, EXAMPLE3, EXAMPLE4, and EXAMPLE5 - none of which will be delivered in a Mastodon-based network (they will be silently dropped).  Other projects are left with two unworkable choices. The mention hack doesn't scale and leaks privacy metadata and users will rebel (they are doing so already). Implementing the spec correctly is the ideal option but as a side effect private messages won't federate to Mastodon. Given those choices we will probably be forced to go with the second. Does anybody have a better way?

**Comment by nightpool** [comment]
> and users will rebel (they are doing so already) 

can you link to what conversations you mean? I would be interested to read actual user feedback in this area

**Comment by redmatrix** [comment]
https://zotadel.net/channel/support/?f=&mid=9ea6f4ca43b308c3fa63cddc7aa32e91fa0063343dc2e07c27042e7cc8b5a68d@zotadel.net

https://grindcore.ch/channel/navigium/?f=&mid=e54312ae5ad9d808d4b9aef21da6daf5981caffa78c894ee57be3d2bbdb0f413@grindcore.ch

Please note that Hubzilla makes extensive use of ACLs. It's part of that project's basic privacy framework. Probably half of the network traffic involves lists of private recipients of varying lengths. Ditto with Diaspora.  

How does Mastodon intend to support privacy groups? I remember this being on the roadmap some time ago.

**Comment by solstag** [comment]
@redmatrix not sure what's up but the first time I click those links they take me to the channel home due to a messed up URL (some rewriting going on), then the second time it works fine.

**Comment by redmatrix** [comment]
My only motivation here is to vote for this issue and provide clear reasoning (basic spec compliance). I'm recommending that Hubzilla go with my second option, which is to implement private ActivityPub messages as per the specification (sans mentions); and consider Mastodon's failure to deliver as an implementation bug. The various project developers can implement workarounds if they can find something acceptable.

**Comment by nightpool** [comment]
yeah that's what I meant by item 2 basically.

**Comment by Gargron** [comment]
I don't really see a way to solve that UX mismatch. The thing is, the person who broadcast to me with a private audience wasn't even someone I followed, so it's not like it could be put into my home timeline. In fact, it's a little bit strange that someone can address things to me in a broadcast when I am not following them.

**Comment by git-marijus** [comment]
This appears to be fixed with #8950
Thanks!
